### ISSUE TITLE
cmd/go: CC variable misparsed when set to a path containing spaces

### ISSUE THREAD

**Issue by XamarinPOC** [golang/go#41400]
go build -buildmode=c-shared -o C:\Users\XAMARI~1\AppData\Local\Temp\gomobile-work-285666087\lib\armeabi-v7a\libbasic.so golang.org/x/mobile/example/basic failed: exit status 2
# runtime/cgo
exec: "C:\\Program": file does not exist

E:\GoWorkDir>go env

set GO111MODULE=auto
set GOARCH=amd64
set GOBIN=
set GOCACHE=C:\Users\Xamarin Developer\AppData\Local\go-build
set GOENV=C:\Users\Xamarin Developer\AppData\Roaming\go\env
set GOEXE=.exe
set GOFLAGS=
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOINSECURE=
set GOMODCACHE=E:\GoWorkDir\pkg\mod
set GONOPROXY=
set GONOSUMDB=
set GOOS=windows
set GOPATH=E:\GoWorkDir
set GOPRIVATE=
set GOPROXY=https://proxy.golang.org,direct
set GOROOT=c:\go
set GOSUMDB=sum.golang.org
set GOTMPDIR=
set GOTOOLDIR=c:\go\pkg\tool\windows_amd64
set GCCGO=gccgo
set AR=ar
set CC=gcc
set CXX=g++
set CGO_ENABLED=1
set GOMOD=
set CGO_CFLAGS=-g -O2
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-g -O2
set CGO_FFLAGS=-g -O2
set CGO_LDFLAGS=-g -O2
set PKG_CONFIG=pkg-config
set GOGCCFLAGS=-m64 -mthreads -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=C:\Users\XAMARI~1\AppData\Local\Temp\go-build823121506=/tmp/go-build -gno-record-gcc-switches

**Comment by davecheney** [comment]
Your go installation is corrupt and/or the time stamps have been disrupted, possibly by unpacking one version of go over another 

Please delete c:\go and reinstall from a fresh copy

**Comment by AlexRouSg** [comment]
I am actually able to repro this.

exact command: `CC="C:\\path with spaces\\bin" go build` or `CC="C:/path with spaces/bin" go build`
works with any source file that uses cgo

It's specific to the `CC/CXX` class of variables so looks like something is wrong with how go parses them.

cc @ianlancetaylor

**Comment by chaitanyaevoke** [comment]
@bcmills I just uninstalled and installed fresh  Go version 1.15.2 but no luck.

**Comment by chaitanyaevoke** [comment]
I'm trying to run build command     => gomobile build -target=android golang.org/x/..............

**Comment by bcmills** [comment]
@AlexRouSg, thanks for the detail. I can even reproduce it on Linux:

```
example.com/bin with spaces$ echo "$CC"
/tmp/tmp.jTseVJjT2L/example.com/bin with spaces/gcc

example.com/bin with spaces$ "$CC" --version
gcc (Debian 9.3.0-13) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


example.com/bin with spaces$ go env CC
/tmp/tmp.jTseVJjT2L/example.com/bin
```

**Comment by AlexRouSg** [comment]
So looks like there are multiple places that needs fixing ...

I've found 
https://github.com/golang/go/blob/29634436fd741a7c685bf8f242b6fd62f093d1ad/src/cmd/cgo/gcc.go#L1536
which only affects running the `cgo` binary direcly

`go build` and `go tool cgo` seems to have a different issue and I can't seem to figure out where that is. But likely it's related to calling `strings.Fields` or something along those lines.

**Comment by bcmills** [comment]
Yes, there are several places where we're calling `strings.Fields` to split this sort of command, which seems clearly incorrect to me.

But I think we also try to follow the usual C toolchain behavior for the CC environment variable, so I'm trying to figure out if there is a standard format for escaping paths containing spaces. (Perhaps @ianlancetaylor or @jayconrod knows?)

**Comment by AlexRouSg** [comment]
@XamarinPOC In the meantime, please use the `PATH` variable to set gcc location instead of using `CC/CXX` if you need to have spaces in the path.

**Comment by XamarinPOC** [comment]
I just uninstalled and installed fresh Go version 1.15.2 but no luck.

I'm trying to run build command => gomobile build -target=android golang.org/x/.............. to generate apk file

**Comment by XamarinPOC** [comment]
@AlexRouSg  Which Sdk?? Can you . please explain it .

**Comment by AlexRouSg** [comment]
@XamarinPOC 
You need to install the android sdk to a path without spaces like `C:\androidsdk`

**Comment by XamarinPOC** [comment]
Yup ,after removing spaces its working now.  Can we get this fix in next release.

**Comment by ianlancetaylor** [comment]
Wait.  Right now I can run

    CC="gcc -O3" go build

and all my cgo files will be built with `-O3`.

Now, maybe that is a mistake and maybe we should change it.  But we definitely shouldn't change that behavior casually.  We should discuss it on the mailing list, we should find out whether it will break anybody.

In particular there is no documented way to pass C compiler flags when running `make.bash` other than setting `CC` to include the compiler and the flags.  And the value of the `CC` environment variable used when running make.bash because the default value of the C compiler used when building programs.

In other words, we can't just treat this a minor bug that we should just fix.  We have to think about how this has worked in the past and how it should work going forward.

**Comment by bcmills** [comment]
The CL that I mailed preserves the property that space-separated arguments following the executable name are treated as arguments: it uses `exec.LookPath` to identify the first space that could plausibly indicate a break between the command and its arguments, and falls back to the old `strings.Fields` behavior if none of the possible prefixes resolves to an executable.

I believe that preserves the existing behavior for all commands that would successfully invoke a compiler, and also fixes some number of cases that previously would have failed.

**Comment by ianlancetaylor** [comment]
OK, thanks.  I'm a bit worried about the possibility of confusion.  I guess it should work but it seems pretty complicated.

**Comment by bcmills** [comment]
Agreed. I was kind of hoping that there would be some existing standard or extension for quoting within `CC`, but I couldn't find one.

As an alternative, I suppose that we could allow shell-style quoting within the CC variable. (Do you think that would be better or worse?)

**Comment by AlexRouSg** [comment]
Shell style quoting sounds better to me. I can imagine people having args like `-I/path with spaces/` which wouldn't be covered by the current CL. Tho that makes me wonder if cgo handles args with spaces correctly for `#cgo CFLAGS` family of comments or pkgconfig provided flags ...

**Comment by ianlancetaylor** [comment]
Shell quoting sounds a little better to me, I guess.

**Comment by jayconrod** [comment]
I'm looking for a narrow fix of #43808 that can be backported, and it will probably affect the eventual solution to this issue.

What I'm thinking right now is that we should support shell quoting using the code in [`go/build.splitQuoted`](https://go.googlesource.com/go/+/refs/tags/go1.16beta1/src/go/build/build.go#1761). That's what we use for `#cgo` directives today, so it seems reasonable to reuse it here (in the absence of any broader standard).

For #43808, we could quote the `CC` and `GCCGO` variables passed from `cmd/go` to `cmd/cgo` if needed, then unquote them in `cmd/cgo`. For 1.17, we could fix this issue by expanding that to `cmd/go` and anywhere else we deal with `CC`. We should fix #43078 at the same time.

**Comment by AlexRouSg** [comment]
@jayconrod 

When I looked at #43808, it seems to me that it would make more sense to not do the path resolution inside `cmd/go` and instead do the resolution within `cmd/cgo` which would avoid setting the `CC` env var and leave behaviour exactly as it was pre security patch.

**Comment by jayconrod** [comment]
@AlexRouSg Hmm, that may work, and it would result in a smaller, less observable change, which would be good for back porting. I'll try that tomorrow morning.

**Comment by seankhliao** [comment]
resolved with #43808 and [CL 284780](https://golang.org/cl/284780)

**Comment by AlexRouSg** [comment]
@seankhliao #43808 only fixes the one specific case of the security fix as it requires backporting, the issue is still unfixed.

**Comment by bcmills** [comment]
Specifically, #43808 fixed the case where `PATH` causes `CC` to be resolved from a directory that contains spaces. This issue is for the case where the `CC` variable itself is an absolute path that includes a space character.

**Comment by justinfx** [comment]
I seem to be hitting this same problem with `CGO_FLAGS` and `CGO_LDFLAGS`, where because it is trying to point at a python installation, it is getting values like:
```
-I"C:/Program Files/Python39/include"
-L"C:/Program Files/Python39"
```
and then ends up splitting "Program" and "Files". Is this a more general problem than only the CC var?

**Comment by bcmills** [comment]
@justinfx, I would not be surprised if the problem is more general than just `CC`, but the fix is probably specific to each variable. I suggest filing a separate issue for `CGO_FLAGS` parsing so that we don't lose track of it.

**Comment by ChristianOConnor** [comment]
I'm having the same problem. I have a little bit of an unconventional setup. I use Windows 10 with Cmder set to bash as admin. I wanted to use the Visual Studio C compiler to build and run a project so I tried `go env -w CC="/mnt/c/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/Llvm/bin/clang-cl.exe"`  and whenever I try to build and run anything using cgo I get this error `cgo: exec C:/tools/Cmder/vendor/git-for-windows/mnt/c/Program: exec: "C:/tools/Cmder/vendor/git-for-windows/mnt/c/Program": file does not exist`  
  
Any workarounds yet?

**Comment by AlexRouSg** [comment]
@ChristianOConnor msvc is not a supported compiler so whether this gets resolved won't help you there.

**Comment by ChristianOConnor** [comment]
> @ChristianOConnor msvc/clang on windows is not supported so whether this gets resolved won't help you there.

Oh okay got you. What is the recommended c compiler for go? MinGW-w64? I'm just worried that this will mess up my build path for compiling C programs in Visual Studio. Is  it safe to have MinGW-w64 installed alongside Visual Studio C build tools?

**Comment by AlexRouSg** [comment]
@ChristianOConnor 

The *only* C compiler on windows is mingw-w64.
> Is it safe to have MinGW-w64 installed alongside Visual Studio C build tools?

As long as you setup your PATH variable correctly it should be fine. Go will look for `gcc`, `g++`, `ar` but you can always use the more specific program name of `CC=x86_64-w64-mingw32-gcc` etc...

**Comment by ChristianOConnor** [comment]
> @ChristianOConnor
> 
> The _only_ (as least the only I could find that is documented to work) C compiler on windows is mingw-w64.
> 
> > Is it safe to have MinGW-w64 installed alongside Visual Studio C build tools?
> 
> As long as you setup your PATH variable correctly it should be fine. Go will look for `gcc`, `g++`, `ar` but you can always use the more specific program name of `CC=x86_64-w64-mingw32-gcc` etc...

Okay awesome thanks!

**Comment by vault-thirteen** [comment]
Why was this issue closed ? 
Was the problem fixed ?

**Comment by ericlagergren** [comment]
Yes, see

> @[gopherbot](https://github.com/gopherbot) gopherbot closed this as [completed](https://github.com/golang/go/issues?q=is%3Aissue+is%3Aclosed+archived%3Afalse+reason%3Acompleted) in https://github.com/golang/go/commit/742dcba7bb953a96c9f3fcdeb32b1c03cbbd8d5e on Aug 16, 2021.
